File: Validacao de Datas/dataCalendario.java
import java.util.Scanner;

class Data {
    short   dia;
    short   mes;
    short   ano;
    boolean ok;
}

public class Main {
    public static int StrToInt(String cadeia) {
        int valor = 0;
        int sinal = 1;
        if (cadeia.isEmpty())
            return 0;
        if (cadeia.charAt(0) == '-') {
            sinal = -1;
            for (int i = 1; i < cadeia.length(); i++) {
                if (!Character.isDigit(cadeia.charAt(i)))
                    return 0;
                valor = valor * 10 + cadeia.charAt(i) - '0';
            }
        } else {
            for (int i = 0; i < cadeia.length(); i++) {
                if (!Character.isDigit(cadeia.charAt(i)))
                    return 0;
                valor = valor * 10 + cadeia.charAt(i) - '0';
            }
        }
        return valor * sinal;
    }

    public static String IntToStr(int valor) {
        return Integer.toString(valor);
    }

    public static String Formatar(int valor, int digitos) {
        String valorCadeia = IntToStr(valor);
        while (valorCadeia.length() < digitos)
            valorCadeia = '0' + valorCadeia;
        return valorCadeia;
    }

    public static boolean anoBissexto(Data data) {
        if (data.ano % 400 == 0)
            return true;
        else if ((data.ano % 4 == 0) && (data.ano % 100 != 0))
            return true;
        else
            return false;
    }

    public static int ultimoDiaMes(Data data) {
        if (data.mes == 2)
            return (anoBissexto(data)) ? 29 : 28;
        if (data.mes <= 7)
            return (data.mes % 2 != 0) ? 31 : 30;
        return (data.mes % 2 != 0) ? 30 : 31;
    }

    public static void validaData(Data data) {
        data.ok = (data.ano >=    1) && 
                  (data.ano <= 9999) &&
                  (data.mes >=    1) && 
                  (data.mes <=   12) &&
                  (data.dia >=    1) && 
                  (data.dia <= ultimoDiaMes(data)) &&
                !((data.ano == 1582) &&
                  (data.mes ==   10) &&
                  (data.dia >=    5) &&
                  (data.dia <=   14));
    }

    public static void alteraData(Data data, int dia, int mes, int ano) {
        data.dia = (short) dia;
        data.mes = (short) mes;
        data.ano = (short) ano;
        validaData(data);
    }

    public static void mostraData(Data data) {
        if (data.ok)
            System.out.println(Formatar(data.dia, 2) + '/' + Formatar(data.mes, 2) + '/' + Formatar(data.ano, 4));
        else
            System.out.println("**/**/****");
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.println("TESTE DE VALIDACAO DE DATA (CALENDARIO)");
        System.out.println("=======================================");
        System.out.println();
        System.out.print("Entre data no formato DD/MM/AAAA: ");

        String DTS = scanner.nextLine();

        int DIA = StrToInt(DTS.substring(0, 2));
        int MES = StrToInt(DTS.substring(3, 5));
        int ANO = StrToInt(DTS.substring(6, 10));

        Data data = new Data();
        alteraData(data, DIA, MES, ANO);

        System.out.print("Data informada: ");
        mostraData(data);

        System.out.println();
        System.out.print("Tecle <Enter> para encerrar... ");
        scanner.nextLine();
    }
}

File: Validacao de Datas/dataCalendario.cpp
 #include <iostream>
#include <string>
using namespace std;

struct Data
{
    short dia;
    short mes;
    short ano;
    bool  ok;
};

int StrToInt(const string& cadeia)
{
    int valor = 0;
    int sinal = 1;
    if (cadeia.empty())
        return 0;
    if (cadeia[0] == '-')
    {
        sinal = -1;
        for (size_t i = 1; i < cadeia.length(); i++)
        {
            if (not isdigit(cadeia[i]))
                return 0;
            valor = valor * 10 + cadeia[i] - '0';
        }
    }
    else
    {
        for (size_t i = 0; i < cadeia.length(); i++)
        {
            if (not isdigit(cadeia[i]))
                return 0;
            valor = valor * 10 + cadeia[i] - '0';
        }
    }
    return valor * sinal;
}

string IntToStr(int valor)
{
    return to_string(valor);
}

string Formatar(int valor, int digitos)
{
    string valorCadeia = IntToStr(valor);
    while (valorCadeia.length() < static_cast<size_t>(digitos))
        valorCadeia = '0' + valorCadeia;
    return valorCadeia;
}

bool anoBissexto(const Data& data)
{
    if (data.ano % 400 == 0)
        return true;
    else if ((data.ano % 4 == 0) and (data.ano % 100 != 0))
        return true;
    else
        return false;
}

int ultimoDiaMes(const Data& data)
{
  if (data.mes == 2) 
    return (anoBissexto(data)) ? 29 : 28;
  if (data.mes <= 7) 
    return (data.mes % 2 != 0) ? 31 : 30;
  return (data.mes % 2 != 0) ? 30 : 31;
}

void validaData(Data& data)
{
    data.ok = (data.ano >=    1) and 
              (data.ano <= 9999) and
              (data.mes >=    1) and 
              (data.mes <=   12) and
              (data.dia >=    1) and 
              (data.dia <= ultimoDiaMes(data)) and
         not ((data.ano == 1582) and
              (data.mes ==   10) and
              (data.dia >=    5) and
              (data.dia <=   14));
}

void alteraData(Data& data, int dia, int mes, int ano)
{
    data.dia = dia;
    data.mes = mes;
    data.ano = ano;
    validaData(data);
}

void mostraData(const Data& data)
{
    if (data.ok)
        cout << Formatar(data.dia, 2) << '/' << Formatar(data.mes, 2) << '/' << Formatar(data.ano, 4) << endl;
    else
        cout << "**/**/****" << endl;
}

int main(void)
{
    cout << "TESTE DE VALIDACAO DE DATA (CALENDARIO)" << endl;
    cout << "=======================================" << endl << endl;
    cout << "Entre data no formato DD/MM/AAAA: ";
    
    string DTS;
    getline(cin, DTS);

    int DIA = StrToInt(DTS.substr(0, 2));
    int MES = StrToInt(DTS.substr(3, 2));
    int ANO = StrToInt(DTS.substr(6, 4));

    Data data;
    alteraData(data, DIA, MES, ANO);

    cout << "Data informada: ";
    mostraData(data);

    cout << endl << "Tecle <Enter> para encerrar... ";
    cin.ignore();

    return 0;
}
